### ISSUE TITLE
Issue with Predator Test Data & Inquiry about Other Datasets (KITTI, ETH) and Test Set Generation

### ISSUE THREAD

**Issue by JefferyXeom** [zhangxy0517/MAC-PLUS-PLUS#3]
Hi, 

Firstly, fantastic work on this point cloud registration project! I've been successfully using your code with the provided test datasets that use FPFH and FCGF features (like 3DMatch/3DLOMatch).

## 1. Problem with the Provided Predator Test Data:

I'm encountering an issue specifically when trying to use the test data associated with Predator features. While the FPFH and FCGF datasets work smoothly with your provided code, the Predator dataset doesn't seem to be compatible out of the box.

Specifics:
When I attempt to run `./Boot --output_path [datasets]/1tok_predator/results/3DLoMatch --input_path [datasets]/1tok_predator/3DLoMatch/all_predator --dataset_name 3dlomatch --descriptor predator --start_index 0`, the program fails.

It seems like the dataloader isn't set up for the Predator feature test set, and after I manually generate the dataload based on your pair description file (those with xx@xx files), the error turns to no .ply files presented.

Could you please clarify how the provided Predator test data is intended to be used with your registration pipeline? Are there specific preprocessing steps or a different script/configuration required for it?

## 2. Inquiry about Other Datasets & Test Set Generation:

I noticed that the currently provided test datasets are primarily 3DMatch and 3DLOMatch. Your paper and code contain tests for datasets like KITTI and ETH

Are there any plans to support or provide pre-processed versions of other common registration datasets like KITTI or ETH in a format compatible with your pipeline?

Alternatively, and perhaps more generally, could you provide some information or (ideally) the scripts/methodology used to generate your existing test datasets (3DMatch, 3DLOMatch with FPFH, FCGF, Predator features) from their raw forms? Understanding your data preparation pipeline would be extremely helpful for users like me who might want to test your method or other outlier removal-focused registration methods on other datasets or custom data.

Any guidance on these points would be greatly appreciated! This would significantly help in benchmarking and extending the use of your excellent registration framework. Besides, your work has been a significant foundation for some of my current research, and I'm exploring avenues to build upon it further. Given my active work in this area, I would be very interested in potentially connecting or discussing your research in more detail if you have the time and inclination.

Thanks for your time and contributions!

Best regards,

Jeffery

**Comment by zhangxy0517** [comment]
Thank you for being so interested in our work.

## 1. Predator Data
All descriptors use the same raw point clouds on the 3DMatch / 3DLoMatch dataset, so you must first copy all `.ply` files and `dataloader.txt` file in each scene of `all_fpfh` to the corresponding scene folder of `all_predator`, then recompile `Boot` from the newest `main.cpp`.

## 2. Other Datasets / Dataloaders
For KITTI and ETH, please refer to [MAC](https://github.com/zhangxy0517/3D-Registration-with-Maximal-Cliques).

To convert the raw data into the format we provide, you can simply save the data loaded by the existing method’s dataloader. Below, we share the code we use to generate the data based on the SC2PCR++ dataloader.
```
import json
import sys
sys.path.append('.')
import argparse
import logging
import open3d as o3d
from tqdm import tqdm
from easydict import EasyDict as edict
from evaluate_metric import TransformationLoss, ClassificationLoss
from dataset import ThreeDLoader
from benchmark_utils import set_seed, icp_refine
from benchmark_utils_predator import *
from utils.timer import Timer
from utils.pointcloud import *
from SC2_PCR import Matcher
set_seed()
from utils.SE3 import *
from collections import defaultdict
import numpy as np

num_node = 'all'
result_path = f'correspondence/{num_node}/'
if not os.path.exists(result_path):
    os.makedirs(result_path)

def eval_3DMatch_scene(loader, matcher, trans_evaluator, cls_evaluator, scene_ind, config):
    num_pair = loader.__len__()
    final_poses = np.zeros([num_pair, 4, 4])

    # 0.success, 1.RE, 2.TE, 3.input inlier number, 4.input inlier ratio,  5. output inlier number
    # 6. output inlier precision, 7. output inlier recall, 8. output inlier F1 score 9. model_time, 10. data_time 11. scene_ind
    stats = np.zeros([num_pair, 12])

    data_timer, model_timer = Timer(), Timer()
    with torch.no_grad():
        error_pair = []
        for i in tqdm(range(num_pair)):
            #################################
            # 1. load data
            #################################
            data_timer.tic()
            scene, src_id, tgt_id, src_keypts, tgt_keypts, src_features, tgt_features, gt_trans = loader.get_data(i)
            folder_path = result_path + scene
            data_time = data_timer.toc()

            if not os.path.exists(folder_path):
                os.makedirs(folder_path)

            #################################
            # 2. match descriptor and compute rigid transformation
            #################################
            model_timer.tic()
            src_keypts, relax_idx, relax_match_points, relax_distance, src_keypts_corr, tgt_keypts_corr = matcher.match_pair(src_keypts, tgt_keypts, src_features, tgt_features)
            model_time = model_timer.toc()

            #################################
            # 3. generate the ground-truth classification result
            #################################
            frag1_warp = transform(src_keypts_corr, gt_trans)
            distance = torch.sum((frag1_warp - tgt_keypts_corr) ** 2, dim=-1) ** 0.5
            gt_labels = (distance < config.inlier_threshold).float()
            corr = torch.cat([src_keypts_corr, tgt_keypts_corr], dim=-1)

            src_keypts = src_keypts[0].data.cpu().numpy()
            tgt_keypts = tgt_keypts[0].data.cpu().numpy()
            relax_idx = relax_idx[0].data.cpu().numpy()
            gt_trans = gt_trans[0].data.cpu().numpy()
            gt_labels = gt_labels[0].data.cpu().numpy()
            corr = corr[0].data.cpu().numpy()
            src_kpt_pcd = make_point_cloud(src_keypts)
            tgt_kpt_pcd = make_point_cloud(tgt_keypts)
            src_name = 'cloud_bin_' + src_id
            tgt_name = 'cloud_bin_' + tgt_id
            src_kpt_path = folder_path + '/' + src_name + '.pcd'
            tgt_kpt_path = folder_path + '/' + tgt_name + '.pcd'
            o3d.io.write_point_cloud(src_kpt_path, src_kpt_pcd)
            o3d.io.write_point_cloud(tgt_kpt_path, tgt_kpt_pcd)
            np.savetxt(folder_path + '/' + src_name + '+' + tgt_name + '@1tok.txt', relax_idx, fmt='%d')
            np.savetxt(folder_path + '/' + src_name + '+' + tgt_name + '@label.txt', gt_labels, fmt='%d')
            np.savetxt(folder_path + '/' + src_name + '+' + tgt_name + '@GTmat.txt', gt_trans, fmt='%.10f')
            np.savetxt(folder_path + '/' + src_name + '+' + tgt_name + '@corr.txt', corr, fmt='%.10f')
            
    return stats, final_poses


def eval_3DMatch(config):
    scene_list = [
        '7-scenes-redkitchen',
        'sun3d-home_at-home_at_scan1_2013_jan_1',
        'sun3d-home_md-home_md_scan9_2012_sep_30',
        'sun3d-hotel_uc-scan3',
        'sun3d-hotel_umd-maryland_hotel1',
        'sun3d-hotel_umd-maryland_hotel3',
        'sun3d-mit_76_studyroom-76-1studyroom2',
        'sun3d-mit_lab_hj-lab_hj_tea_nov_2_2012_scan1_erika'
    ]
    all_stats = {}
    for scene_ind, scene in enumerate(scene_list):
        loader = ThreeDLoader(root=config.data_path,
                                descriptor=config.descriptor,
                                inlier_threshold=config.inlier_threshold,
                                num_node=config.num_node,
                                use_mutual=config.use_mutual,
                                select_scene=scene,
                                )
        matcher = Matcher(inlier_threshold=config.inlier_threshold,
                            num_node=config.num_node,
                            use_mutual=config.use_mutual,
                            d_thre=config.d_thre,
                            num_iterations=config.num_iterations,
                            ratio=config.ratio,
                            nms_radius=config.nms_radius,
                            max_points=config.max_points,
                            k1=config.k1,
                            k2=config.k2, 
                            FS_TCD_thre=config.FS_TCD_thre,
                            relax_match_num=config.relax_match_num,
                            NS_by_IC=config.NS_by_IC)
        trans_evaluator = TransformationLoss(re_thre=config.re_thre, te_thre=config.te_thre)
        cls_evaluator = ClassificationLoss()

        scene_stats = eval_3DMatch_scene(loader, matcher, trans_evaluator, cls_evaluator, scene_ind, config)
        all_stats[scene] = scene_stats
    logging.info(f"Max memory allicated: {torch.cuda.max_memory_allocated() / 1024 ** 3:.2f}GB")



if __name__ == '__main__':
    from config import str2bool

    parser = argparse.ArgumentParser()
    parser.add_argument('--config_path', default='', type=str, help='snapshot dir')
    parser.add_argument('--solver', default='SVD', type=str, choices=['SVD', 'RANSAC'])
    parser.add_argument('--use_icp', default=False, type=str2bool)
    parser.add_argument('--save_npy', default=True, type=str2bool)
    args = parser.parse_args()

    config_path = args.config_path
    config = json.load(open(config_path, 'r'))
    config = edict(config)

    import os
    os.environ['CUDA_VISIBLE_DEVICES'] = config.CUDA_Devices
    if not os.path.exists("./logs"):
        os.makedirs("./logs")

    log_filename = f'logs/3DLoMatch-{config.descriptor}.log'
    logging.basicConfig(level=logging.INFO,
                        filename=log_filename,
                        filemode='a',
                        format="")
    logging.getLogger().addHandler(logging.StreamHandler(sys.stdout))

    # evaluate on the test set
    stats = eval_3DMatch(config)
```

**Comment by JefferyXeom** [comment]
Thanks for your rapid and explict reply. Things work fine now😊.
